Implement LLVM IR Virtual Function Elimination for Swift classes.

kubamracek opened the pull request:
- Virtual calls are done via a @llvm.type.checked.load instrinsic call with a type identifier
- Type identifier of a vfunc is the base method's mangling
- Type descriptors and class metadata get !type markers that list offsets and type identifiers of all vfuncs
- The -enable-llvm-vfe frontend flag enables VFE
- Two added tests verify the behavior on IR and by executing a program

Depends on changes in LLVM (some unmerged yet): https://reviews.llvm.org/D107645, https://reviews.llvm.org/D108741, https://reviews.llvm.org/D109114, https://reviews.llvm.org/D109169.

eeckstein commented on lib/IRGen/GenMeta.cpp:
```diff
@@ -334,6 +334,33 @@ void IRGenModule::emitNonoverriddenMethodDescriptor(const SILVTable *VTable,
+    var->addTypeMetadata(offset, typeIdForMethod(*this, method));
+  }
+
+  switch (getOptions().VCallVisibility) {
```
Why not deriving the VCall visibility from the method visibility?
Setting the visibility for _all_ calls sounds dangerous, e.g. if a public method gets a "File" visibility.

eeckstein replied:
You can call a private/internal method from a function, annotated with `@_optimize(none)`. That will keep the virtual call alive

kubamracek replied:
Changed the PR to infer the vtable visibility from the class decl's access scope visibility.

kubamracek commented on lib/IRGen/GenMeta.cpp:
```diff
@@ -334,6 +334,33 @@ void IRGenModule::emitNonoverriddenMethodDescriptor(const SILVTable *VTable,
 
+void IRGenModule::addVTableTypeMetadata(llvm::GlobalVariable *var,
+    SmallVector<std::pair<Size, SILDeclRef>, 8> vtableEntries) {
+  if (!getOptions().VirtualFunctionElimination)
```
Removed.

kubamracek committed:
Implement LLVM IR Virtual Function Elimination for Swift classes.

- Virtual calls are done via a @llvm.type.checked.load instrinsic call with a type identifier
- Type identifier of a vfunc is the base method's mangling
- Type descriptors and class metadata get !type markers that list offsets and type identifiers of all vfuncs
- The -enable-llvm-vfe frontend flag enables VFE
- Two added tests verify the behavior on IR and by executing a program

eeckstein reviewed (approved):
LGTM.
It would be good if @aschwaighofer takes a look, too.

aschwaighofer reviewed (approved):
lgtm

kubamracek commented:
@swift-ci please test

kubamracek commented:
@swift-ci please test

kubamracek commented:
@swift-ci please test

kubamracek commented:
@swift-ci please test

kubamracek commented:
@swift-ci please test
